libnavigation-core/src/main/java/com/mapbox/navigation/core/RoutesRefreshDataProvider.kt

```diff
@@ -0,0 +1,39 @@
+package com.mapbox.navigation.core
+
+import com.mapbox.navigation.base.route.NavigationRoute
+
+internal data class RoutesRefreshData(
+    val primaryRoute: NavigationRoute,
+    val primaryRouteProgressData: RouteProgressData,
+    val alternativeRoutesProgressData: List<Pair<NavigationRoute, RouteProgressData?>>
+) {
+    val allRoutesRefreshData = listOf(primaryRoute to primaryRouteProgressData) +
+        alternativeRoutesProgressData
+}
+
+internal class RoutesRefreshDataProvider(
+    private val routesProgressDataProvider: RoutesProgressDataProvider,
+) {
+
+    /**
+     * Retrieved progress data for passed routes.
+     *
+     * @throws IllegalArgumentException if routes are empty
+     */
+    @Throws(IllegalArgumentException::class)
+    suspend fun getRoutesRefreshData(
+        routes: List<NavigationRoute>
+    ): RoutesRefreshData {
+        if (routes.isEmpty()) {
+            throw IllegalArgumentException("Routes must not be empty")
+        }
+        val routesProgressData = routesProgressDataProvider
+            .getRouteRefreshRequestDataOrWait()
+        val primary = routes.first()
+        val alternatives = routes.drop(1)
+        val alternativeRouteProgressDatas = alternatives.map { route ->
+            route to routesProgressData.alternatives[route.id]
```

Comment:
What do you mean? We get route progress once we want to refresh the route. That's the best we can do, what chance of desync do you see here?

Reply:
I used to think that map will throw an exception if there is no value for a key. But it returns null, so even in case of desync the SDK does the best it can.